File: effects/bouncing_balls.lua
name = "Bouncing Balls"

local DEFAULT_COLORS = {{255, 0, 0}, {0, 255, 0}, {0, 0, 255}}

local balls = {}

function init(n)
    balls = {}
    for i, c in ipairs(DEFAULT_COLORS) do
        table.insert(balls, {
            r         = c[1],
            g         = c[2],
            b         = c[3],
            height    = 1.0,
            velocity  = -(i * 0.5),
            dampening = 0.85 + math.min(0.1, (i - 1) * 0.02),
        })
    end
end

function update(buf, dt)
    local n       = buf:len()
    local gravity = -9.8 * n * 0.01

    buf:clear()

    for _, ball in ipairs(balls) do
        ball.velocity = ball.velocity + gravity * dt
        ball.height   = ball.height   + ball.velocity * dt

        if ball.height <= 0.0 then
            ball.height   = 0.0
            ball.velocity = -ball.velocity * ball.dampening
            if math.abs(ball.velocity) < 0.01 then
                ball.velocity = 1.0
            end
        end

        ball.height = math.max(0.0, math.min(1.0, ball.height))
        local pos   = (1.0 - ball.height) * (n - 1)
        buf:plot(pos, ball.r, ball.g, ball.b, 1.0)
    end

    return true
end

File: src/effects/bouncing_balls.rs
use std::time::Duration;
use crate::effects::{Effect, Buffer, plot};
use crate::pixels::Color;

const DEFAULT_COLORS: &[Color] = &[
    [255, 0, 0],
    [0, 255, 0],
    [0, 0, 255],
    [255, 255, 0],
    [0, 255, 255],
    [255, 0, 255],
];

struct Ball {
    color: Color,
    height: f32,
    velocity: f32,
    dampening: f32,
}

pub struct BouncingBallsEffect {
    balls: Vec<Ball>,
}

impl BouncingBallsEffect {
    pub fn new(colors: Vec<Color>, ball_count: usize) -> Self {
        let ball_colors: Vec<Color> = if colors.is_empty() {
            let n = ball_count.min(DEFAULT_COLORS.len());
            DEFAULT_COLORS[..n].to_vec()
        } else {
            colors
        };

        let balls = ball_colors
            .into_iter()
            .enumerate()
            .map(|(i, color)| Ball {
                color,
                height: 1.0,
                velocity: -((i as f32 + 1.0) * 0.5),
                dampening: 0.85 + (i as f32 * 0.02).min(0.1),
            })
            .collect();

        Self { balls }
    }
}

impl Effect for BouncingBallsEffect {
    fn name(&self) -> &'static str { "Bouncing Balls" }

    fn update(&mut self, buffer: &mut Buffer, delta: Duration) -> bool {
        let n = buffer.len();
        let dt = delta.as_secs_f32();
        // gravity scaled to strip length so a ball takes ~1s to fall the full strip
        let gravity = -9.8 * n as f32 * 0.01;

        for pixel in buffer.iter_mut() {
            *pixel = [0, 0, 0];
        }

        for ball in self.balls.iter_mut() {
            ball.velocity += gravity * dt;
            ball.height += ball.velocity * dt;

            if ball.height <= 0.0 {
                ball.height = 0.0;
                ball.velocity = -ball.velocity * ball.dampening;
                // prevent balls from getting stuck due to float precision
                if ball.velocity.abs() < 0.01 {
                    ball.velocity = 1.0;
                }
            }

            ball.height = ball.height.clamp(0.0, 1.0);
            let pos = (1.0 - ball.height) * (n - 1) as f32;
            plot(buffer, pos, ball.color, 1.0);
        }

        true
    }
}
